### ISSUE TITLE
Provide a mechanism to easily pre-create new indices before a rollover period occurs using randomisation to spread creates out over a specific time range

### ISSUE THREAD

**Issue by geekpete** [elastic/curator#1319]
The use case is where rollover triggers a vast amount of indices to be created at once.
This simultaneous index creation can sometimes affect cluster in negative ways if there's enough indices being created at once.

If there was a way to pre-create indices for soon to be rolled over index patterns and spread those creates out then this impact might be mitigated.

Even if the method to splay the index creates is just randomised delays between the create for each index pattern for 1-30 seconds perhaps, with those settings configurable. 
There would also need to be some configuration to say how far ahead of rollover the creates need to occur, but this might just need to ensure the next period's index is always precreated and ready so ensure there's always now+1 index created when it checks on each run.

Also wondering what the downsides to such an approach might be.

**Comment by geekpete** [comment]
Yes new index creation at UTC for daily or whatever period indices.

So if you just had a wait for green and created indices sequentially for existing create index functionality, this might mean that some events timestamped for the next period land in the previous period's index if there turns out to be enough delay in the index creation for whatever reason (very large list of indices or loaded cluster making each create a non-neglible delay), a situation you'd want to avoid?

The sequential and wait for green option would work if it was used with creating n+1 period indices to ensure index creation was always ahead by one period and the current index receiving ingest is changed by just being flipped to the next index that was already precreated from a previous curator run.

So the point around Beats/Logstash and ILM was that indices might roll over more frequently and not just at UTC time, so would dilute the create impact on the cluster more?

Depending on the difficulty of adding n+1 precreate logic combined with wait for green+sequential functionality to the existing create_index code, there has to be a weighing up of the usefulness and need for this feature vs the amount of time before most deployments are using the newer ILM stuff.

Support will exist for the pre-ILM versions for a while so anyone needing this feature due to being on pre-ILM capable stacks could certainly benefit from it.
The other side of the coin is how much of an edge case this scenario is and how many users it might affect, I've seen it a handful of times now anecdotally.
The ability to write a script to pre-create indices might also be fairly simple for the edge case users to do themselves as well, but it'd be a nice to have feature built into Curator.

**Comment by geekpete** [comment]
I was thinking about scenarios where you have something indexing against an alias and are creating indices sequentially before switching the alias and any delay causes that alias to stay on the previous index past the UTC rollover time then you'd have events in the wrong index potentially.
Rollover Index API does both the create new index and move alias steps in one go and these cannot be separated so that wouldn't happen there. Logstash indexing to a dynamic index name with the period in the index name will create a missing index for the current period if UTC rollover has occurred, so wouldn't occur there. This risk would only be for a workflow that has these steps separated, such as a Curator recipe that has index create actions followed by dependent alias switch actions.

If you're creating many indices at UTC rollover and are seeing some impact from that, then creating 10k indices concurrently but ahead of time still has the same impact. What matters is mitigating that potential impact (via sequential and wait for green or whatever mechanism), so it's not only creating ahead of time but doing so with a strategy for reduced impact.

The purpose of this feature might already be mostly achievable with existing capability.
If you're creating 100 indices at UTC roll over time, then you could use index create actions and alias actions after that.

Index create already has date math to be able to create tomorrow's indices ahead of time:
https://www.elastic.co/guide/en/elasticsearch/client/curator/current/create_index.html#_date_math

The only question is then how do you avoid creating all the indices at once.
You could just split/group the index creations into separate curation runs and use the cron `RANDOM_DELAY` environment variable to add splay or add randomization to your script execution directly with a preceding delay using scripting or just space them out manually to avoid overlapping.


So there's a bunch of ways to achieve it and I think you might be right in suggesting that future direction of ILM (Index Lifecycle Management) might largely mitigate this type of impact, but I think the wait for green might be handy addition to create_index perhaps. I guess you'll hit the timeout if you only get to yellow for whatever reason (eg replicas)?

**Comment by MarkStoneLin** [comment]
I think it‘s’ very necessary for users to use pre-create, especially when huge data come in 00:00.
The ES warning The operations group in ervey 00:00, because the automatic creation in 00:00, but the data DO NOT wait the automatic creation, then the data congestion happen. 
I really hope someone can share his experiences about solving this problem, or at least updata the DATE MATH, to support more formats of date, not only can format 2017.3.28, but also can 20170328 or others.
